src/pprof/ddprof_pprof.cc

```diff
@@ -34,54 +32,77 @@ std_string_2_slice_c_char(const std::string &str) {
-    perf_value_type[value_index].unit = perf_unit;
+DDRes pprof_create_profile(DDProfPProf *pprof, DDProfContext *ctx) {
+  // Figure out which sample_type_ids are used by active watchers
+  const PerfWatcher *watchers = ctx->watchers;
+  size_t num_watchers = ctx->num_watchers;
+  bool active_ids[DDPROF_PWT_LENGTH] = {};
+  struct ddprof_ffi_ValueType perf_value_type[DDPROF_PWT_LENGTH];
+  for (unsigned i = 0; i < num_watchers; ++i) {
+    int this_id = watchers[i].sample_type_id;
+    if (this_id < 0 || this_id == DDPROF_PWT_NOCOUNT || this_id >= DDPROF_PWT_LENGTH) {
+      LG_WRN("Watcher \"%s\" (%d) has invalid sample_type_id %d, ignoring",
+          watchers[i].desc, i, this_id);
+      continue;
+    }
+    active_ids[watchers[i].sample_type_id] = true;
+  }
+
+  // Now that we have all of the sample ids, we populate the sample_type_id 
+  // permutation vector. This allows the exporter to take the absolute sample ID 
+  // and convert it into the index of the pprof.
+  memset(ctx->sample_type_pv, 0, sizeof(*ctx->sample_type_pv)*DDPROF_PWT_LENGTH);
+  unsigned num_sample_type_ids = 0;
+  for (unsigned i = 0; i < DDPROF_PWT_LENGTH; ++i) {
+    if (!active_ids[i])
+      continue;
+    const char *value_name = profile_name_from_idx(i);
+    const char *value_unit = profile_unit_from_idx(i);
+    if (!value_name || !value_unit) {
+      LG_WRN("Malformed sample type (%d), ignoring", i);
+      continue;
+    }
+    ctx->sample_type_pv[i] = num_sample_type_ids;
+    perf_value_type[num_sample_type_ids].type_ = (ddprof_ffi_Slice_c_char) {
+      .ptr = value_name,
+      .len = strlen(value_name)};
+    perf_value_type[num_sample_type_ids].unit = (ddprof_ffi_Slice_c_char) {
+      .ptr = value_unit,
+      .len = strlen(value_unit)};
+    ++num_sample_type_ids;
+  }
+
+  // If none of the samples were good, that's an error
+  if (!num_sample_type_ids) {
+    // We use the phrase "profile type" in the error, since this is more
+    // obvious for customers.
+    DDRES_RETURN_ERROR_LOG(DD_WHAT_PPROF, "No valid profile types given");
   }
 
-  pprof->_nb_values = value_index;
+  pprof->_nb_values = num_sample_type_ids;
   struct ddprof_ffi_Slice_value_type sample_types = {.ptr = perf_value_type,
-                                                     .len = value_index};
-  struct ddprof_ffi_Period period;
-  if (options->freq) {
-    period.type_ = s_cpu_time;
-    period.value = (s_nanos_in_one_sec) / options->sample_frequency;
-  } else {
-    period.type_ = s_cpu_sample;
-    period.value = options->sample_period;
+                                                     .len = pprof->_nb_values};
+
+  // To populate the defaults, we take the period/frequency of the first watcher
+  // with the first valid sample_type_id.  The checks above ensure we have one
+  // such watcher.
+  int64_t periodfreq_default = 0;
+  for (unsigned i = 0; i < num_watchers; ++i) {
+    if (ctx->sample_type_pv[0] == watchers[i].sample_type_id) {
```

Comment:
I think the right test here is `ctx->sample_type_pv[watchers[i].sample_type_id] >= 0` (if `sample_type_id` is initialized to -1).

Reply:
Yes, this test was totally broken. This is somewhat irrelevant in the new code though.